includes/class-events.php

```diff
@@ -89,56 +91,41 @@ public function populate_concurrent_action_whitelist() {
 	}
 
 	/**
-	 * List events pending for the current period
+	 * List events pending for the current period.
 	 *
-	 * @param array $job_queue_size   Maximum number of events to return (excludes internal events).
-	 * @param array $job_queue_window How many seconds into the future events should be fetched.
-	 * @return array
+	 * @param null|int $job_queue_size   Maximum number of events to return (excludes internal events).
+	 * @param null|int $job_queue_window How many seconds into the future events should be fetched.
+	 * @return array Events to be run in the next batch.
 	 */
-	public function get_events( $job_queue_size = null, $job_queue_window = null ) {
+	public function get_events( $job_queue_size = null, $job_queue_window = null ): array {
 		$job_queue_size   = is_null( $job_queue_size ) ? JOB_QUEUE_SIZE : $job_queue_size;
 		$job_queue_window = is_null( $job_queue_window ) ? JOB_QUEUE_WINDOW_IN_SECONDS : $job_queue_window;
 
-		$events = get_option( 'cron' );
+		// Grab relevant events that are due, or soon will be.
+		$current_time = time();
+		$events = self::query( [
+			'timestamp' => [ 'from' => 0, 'to' => $current_time + $job_queue_window ],
+			'status'    => Events_Store::STATUS_PENDING,
+			'limit'     => -1, // Need to get all, to ensure we grab internals even when queue is backed up.
+		] );
 
 		// That was easy.
-		if ( ! is_array( $events ) || empty( $events ) ) {
-			return array(
-				'events' => null,
-			);
+		if ( empty( $events ) ) {
+			return [ 'events' => null ];
 		}
 
-		// Simplify array format for further processing.
-		$events = collapse_events_array( $events );
-
-		// Select only those events to run in the next sixty seconds.
-		// Will include missed events as well.
-		$current_events  = array();
-		$internal_events = array();
-		$current_window  = strtotime( sprintf( '+%d seconds', $job_queue_window ) );
-
+		$current_events  = [];
+		$internal_events = [];
 		foreach ( $events as $event ) {
-			// Skip events whose time hasn't come.
-			if ( $event['timestamp'] > $current_window ) {
-				continue;
-			}
-
-			// Skip events that don't have any callbacks hooked to their actions, unless their execution is requested.
-			if ( ! $this->action_has_callback_or_should_run_anyway( $event ) ) {
-				continue;
-			}
```

Comment:
In theory this feature makes sense, but it never worked in practice and we ended up disabling it via filter in mu-plugins b/c it would cause runaway event creation (recurring ones would just keep being recreated over and over).

So I'm just removing it altogether. I think the safer option will be to just throw a PHP warning when an event runs but has no callback.